### DISCUSSION TITLE
Capturing all text entry including keystroke timing

### DISCUSSION THREAD

**Discussion by angoodkind** [jspsych/jsPsych#1539] [Q&A] [answered]
I’m not sure why I’m having so much trouble with this, but I’m just trying to capture the text and keystroke information in a free text entry field. For example, if a user typed “I am”, the following would be captured:

```
Text: "I am"
Keystroke, Keypress, KeyUp
SHIFT, 1, 3
i, 2, 3
SPACE, 4, 5
a, 6, 7
m, 8, 9
```
Is this straightforward to write? (Eventually I would like to display an image while they complete each trial.)

Thanks,
Adam

**Comment by jodeleeuw** [comment]
Hi Adam,

At the moment none of the included plugins will record this kind of info. You could modify the survey-text plugin to do this, or create a plugin from scratch. I think it will be relatively straightforward to write if/once you have an understanding of keyboard events in JavaScript. [This tutorial](https://www.javascripttutorial.net/javascript-dom/javascript-keyboard-events/) looks pretty clear to me.

This is an increasingly common request that we are getting, so if you do put together a plugin to do this let us know and consider submitting a pull request!

**Reply by angoodkind** [reply]
Thanks so much Josh. I've been talking to Matt Crump, and he pointed me to an older plugin that he (and you?) had worked on: [html-keyboard-multiple-response](https://github.com/CrumpLab/jspsychrexamples/blob/master/Stroop_multiple_response/experiment/jspsych-6/plugins/jspsych-html-keyboard-multiple-response.js). My plan is to combine it with `image-keyboard-response`, since my stimuli will be pictures. 

I think these two plugins have (almost) everything I need, and will be happy to submit my plugin when it's complete.

EDIT: I feel foolish; I believe Matt just left your name on there since you wrote the original plugin.

**Reply by angoodkind** [reply]
@jodeleeuw - I have a working plugin, but still need to do a lot of formatting work. Here is an MWE, though:
![Screen Shot 2021-02-20 at 1 18 54 PM](https://user-images.githubusercontent.com/1312344/108606224-71f6b800-737e-11eb-8c18-652f54fde227.png)

Which produces this output:
```
[
	{
		"rt": [
			1546,
			1619,
			2160,
			2444,
			2609,
			2939,
			3254,
			3569,
			3689,
			4053.0000000000005,
			4383.000000000001,
			4592,
			4713,
			4952,
			5133.000000000001,
			5433,
			5642,
			5882,
			6078,
			6557,
			6769,
			7097,
			7381,
			7620.000000000001,
			7818,
			8521,
			13635.000000000002
		],
		"stimulus": "img/green-blue-bus.png",
		"key_press": [
			16,
			73,
			32,
			65,
			77,
			32,
			84,
			89,
			80,
			73,
			78,
			71,
			32,
			84,
			72,
			73,
			83,
			32,
			83,
			69,
			78,
			84,
			69,
			78,
			67,
			69,
			91
		],
		"key_name": [
			"\u0010",
			"I",
			" ",
			"A",
			"M",
			" ",
			"T",
			"Y",
			"P",
			"I",
			"N",
			"G",
			" ",
			"T",
			"H",
			"I",
			"S",
			" ",
			"S",
			"E",
			"N",
			"T",
			"E",
			"N",
			"C",
			"E",
			"["
		],
		"color": "green-blue",
		"trial_type": "image-keyboard-multiple-response",
		"trial_index": 0,
		"time_elapsed": 116705,
		"internal_node_id": "0.0-0.0-0.0"
	}
]
```

**Comment by angoodkind** [comment, marked as answer]
I have resolved this issue.
